routers/repo/http.go

```diff
@@ -95,6 +95,13 @@ func HTTP(ctx *context.Context) {
 	}
 
 	repo, err := models.GetRepositoryByName(repoUser.ID, reponame)
+	if models.IsErrRepoNotExist(err) && setting.Repository.Upload.AutoCreate {
```

Comment:
This doesn't look good to me. You should check err for each function. Later err check block is expecting only `GetRepositoryByName` error, not `CreateRepository` error. I think something like this:
```
	repo, err := models.GetRepositoryByName(repoUser.ID, reponame)
	if err != nil {
		if models.IsErrRepoNotExist(err) {
			if setting.Repository.Upload.AutoCreate {
				repo, err = models.CreateRepository(repoUser, repoUser, models.CreateRepoOptions{
					Name: reponame,
				})
				if err != nil {
					ctx.Handle(http.StatusInternalServerError, "CreateRepository", err)
					return
				}
			} else {
				ctx.Handle(http.StatusNotFound, "GetRepositoryByName", nil)
				return
			}
		} else {
			ctx.Handle(http.StatusInternalServerError, "GetRepositoryByName", err)
			return
		}
	}
```

Reply:
Also you are creating repository on any action, you should check if user is doing push or pull. I think you can use `accessMode` variable defined few lines above (changing if condition to `accessMode == models.AccessModeWrite && setting.Repository.Upload.AutoCreate`